Improved Chat Component API

Mystiflow opened the pull request:
See description of every commit for further detail.

- Shared code for various BaseComponent toLegacyText and toPlainText methods moved to the superclass to cut back on duplicate code and make it more accessible
- Various unit tests inserted to accommodate new changes
--
**ComponentBuilder:**
- No arg constructor. Allows for creating without an initial component (e.g. "" TextComponent)
- Better control of parts (get at index, remove at index, get the most recent part -getCurrentComponent)
- Cursor API. For example; when there are two components in a builder, setting the cursor to 0 and then using ```setColor(ChatColor.BLUE)``` will set the color of the first component instead of the second by default. Or to change the previous part ```setCursor(getCursor() - 1).setColor(ChatColor.BLUE)```
--

Janmm14 commented on chat/src/test/java/net/md_5/bungee/api/chat/ComponentsTest.java:
```diff
@@ -47,7 +86,7 @@ public void testComponentFormatRetention()
     public void testBuilderClone()
     {
-        ComponentBuilder builder = new ComponentBuilder( "Hel" ).color( ChatColor.RED ).append( "lo" ).color( ChatColor.DARK_RED );
+        ComponentBuilder builder = new ComponentBuilder( "Hello " ).color( ChatColor.RED ).append( "world" ).color( ChatColor.DARK_RED );
```
lul why change it at all?

Mystiflow replied:
OCD.. 🤪

md-5 commented:
My only question is what's the purpose of the dummy component?

kamcio96 commented on chat/src/main/java/net/md_5/bungee/api/chat/ComponentBuilder.java:
```diff
@@ -88,11 +146,18 @@ public ComponentBuilder append(BaseComponent component)
-        BaseComponent previous = current;
-        current = component.duplicate();
-        current.copyFormatting( previous, retention, false );
+        BaseComponent previous = parts.get( parts.size() - 1 );
```
Can throw exception when created with no args constructor

kamcio96 commented:
@md-5 you can set format before appending text

Mystiflow commented:
Yes, the dummy component is for setting formatting before an appendment has been made like kamcio has said.

@kamcio96 
I have added your constructor suggestion and fixed the IllegalArgumentException. I have also added some more unit tests. If you have any ideas for what else may need to be tested, let me know

kamcio96 commented on chat/src/test/java/net/md_5/bungee/api/chat/ComponentsTest.java:
```diff
@@ -11,24 +11,120 @@
+    public void testComponentGettingExceptions()
+    {
+        ComponentBuilder builder = new ComponentBuilder();
+        builder.getComponent( -1 );
```
Test fails here, you should rewrite it to few methods.

Mystiflow replied:
The test passes for me, it should expect for IndexOutOfBoundsException

md-5 commented:
Tests pass for me also.

Looks good to me. Any other comments @kamcio96 @Janmm14  ?
